src/Symfony/Component/Routing/RouteCompiler.php

```diff
@@ -168,10 +196,21 @@ private static function compilePattern(Route $route, $pattern, $isHost)
         for ($i = 0, $nbToken = count($tokens); $i < $nbToken; ++$i) {
             $regexp .= self::computeRegexp($tokens, $i, $firstOptional);
         }
+        $regexp = self::REGEX_DELIMITER.'^'.$regexp.'$'.self::REGEX_DELIMITER.'s'.($isHost ? 'i' : '');
+
+        // enable Utf8 matching if really required
+        if ($needsUtf8) {
+            $regexp .= 'u';
+            for ($i = count($tokens) - 1; $i >= 0; --$i) {
```

Comment:
why not just iterating over tokens ? (we don't actually need the reverse ordering)